src/pubsub.c

```diff
@@ -334,25 +305,15 @@ int pubsubUnsubscribeChannel(client *c, robj *channel, int notify, pubsubtype ty
         if (server.cluster_enabled && type.shard) {
             slot = getKeySlot(channel->ptr);
         }
-        d = *type.serverPubSubChannels(slot);
-        serverAssertWithInfo(c,NULL,d != NULL);
-        de = dictFind(d, channel);
+        de = kvstoreDictFind(*type.serverPubSubChannels, slot, channel);
         serverAssertWithInfo(c,NULL,de != NULL);
         clients = dictGetVal(de);
         serverAssertWithInfo(c, NULL, dictDelete(clients, c) == DICT_OK);
         if (dictSize(clients) == 0) {
             /* Free the dict and associated hash entry at all if this was
              * the latest client, so that it will be possible to abuse
              * Redis PUBSUB creating millions of channels. */
-            dictDelete(d, channel);
-            if (type.shard) {
-                if (dictSize(d) == 0) {
-                    dictRelease(d);
-                    dict **d_ptr = type.serverPubSubChannels(slot);
-                    *d_ptr = NULL;
-                }
```

Comment:
not sure what you mean by "regression" in this case

correct, kvstore's dicts always exist, even if empty

Reply:
i mean that the previous code dismissed the unused dicts, and now they always exist.
maybe we should modify kvstore to be able to dismiss empty dicts?
maybe we don't want that in the main db keys, but we do want that in pubsub?
@madolson WDYT?

Reply:
Yeah, we did that for the pubsub dictionaries to avoid the unnecessary overhead when it wasn't being used (since we expect many users wouldn't use it at all).

Reply:
so what's the verdict? do we want `kvstore` to deallocate empty dicts?

if so, why shouldn't we do it in the main `kvstore`? i'm assuming that if clustering mode is used, there's probably more than one shard, and assuming fair distribution, maybe it's safe to assume that >0/5 of the hash slots are unused in every shard.

Reply:
yes and no. there are few concerns:
1. if sharded pubsub is completely unused, let's not allocate all these dicts.
2. for the sharded keyspace, indeed we'd rather not allocate dicts for slots that aren't assigned to the current node.
3. but on the other hand, we'd rather not de-allocate a dict every time it becomes empty, and re-allocate it when a key is added again.

maybe the easiest way forward is to add a per-kvstore flag, so that pubsub can behave differently than the keyspace ones?

Reply:
do we want to trouble ourselves with constant memory overheads? same goes for that other 200k overhead

Reply:
this is 1mb, not 200k, i.e. in case sharded pubsub isn't used at all.
and it's another ~2mb on shards that don't have a lot of slots (for keys and expires).
i do think we should make an effort to reduce there. i can imagine some cases with many small shards in which it'll cause a significant impact.
i don't think this PR should bother to make a distinction between slots that are unassigned, and ones that are simply empty. but it is quite easy to avoid regression compared to what's before this PR, and avoid keeping empty pubsub dicts (1mb), and avoid the extra per-dict kvstore pointer in https://github.com/redis/redis/pull/12822#discussion_r1457839748, so let's do that.